File: AtoI.py
class Solution:
    def myAtoi(self, s: str) -> int:
        length = len(s)
        number = 0
        count=0
        neg = False
        pos = False
        first_int=False
        s = list(s)
        for i in range(length):
            if s[i] == " ":
                if first_int==True:
                    break
                elif neg == True or pos == True:
                    break
                else:
                    continue
            elif s[i].isdigit():
                if number == 0:
                    number = int(s[i])
                    first_int=True
                else:
                    number = (number*(10)) + int(s[i])
                    count+=1
                    if count == 1:
                        first_int=True
            elif s[i]=='-':
                if first_int==True:
                    break
                elif neg == True:
                    break
                else:
                    neg = True
            elif s[i]=='+':
                if first_int==True:
                    break
                elif pos == True:
                    break
                else:
                    pos = True
                    continue
            elif (s[i].isdigit() == False):
                break
            else: 
                continue
        if neg == True:
            if neg == True and pos == True:
                number = 0
            else:
                number=number*(-1)
        if (number<(-2**31)):
            number = -2**31
        elif (number>((2**31)-1)):
            number = (2**31)-1
        return number

File: AtoI.cpp
class Solution {
public:
    int myAtoi(string s) {
        int i = 0;
        long long res = 0;
        bool num = false;
        bool is_neg = false;
        while(i<s.size())
        {
            // allowing whitespace
            if(!num && s[i] == ' ')
            {
                i++;
                continue;
            }
            // allowing negation
            else if(!num && s[i] == '-' && i<s.size() && isdigit(s[i+1]))
            {
                is_neg = true;
            }
            // allowing positive
            else if(!num && s[i] == '+' && i<s.size() && isdigit(s[i+1]))
            {
                is_neg = false;
            }
            // scanning for nums
            else if(isdigit(s[i]))
            {
                num = true;
                int digit = s[i] - '0';

                if (res > ((INT_MAX - digit) / 10) && !is_neg)
                    return INT_MAX;
                
                else if (res > (((1LL*INT_MAX+1) - digit) / 10) && is_neg)
                    return INT_MIN;

                res = res * 10 + digit;
            }
            // otherwise breaking from the loop
            else
            {
                break;
            }
            i++;
        }
        // negation if true
        if(is_neg)
        {
            res = -res;
        }
        // capping the value
        if (res > INT_MAX)
            return INT_MAX;

        if (res < INT_MIN)
            return INT_MIN;

        return (int)res;
    }
};